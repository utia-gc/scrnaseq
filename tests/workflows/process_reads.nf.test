nextflow_workflow {

    name "Test Workflow PROCESS_READS"
    script "workflows/process_reads.nf"
    workflow "PROCESS_READS"

    test("PROCESS_READS concatenates and trims reads by default -- SE reads.") {

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [
                            sampleName:   "SRR1066657",
                            sampleNumber: "3",
                            lane:         "001",
                            readType:     "single"
                        ],
                        [
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR1066657_S3_L001_R1_001.fastq.gz")
                        ]
                    ],
                    [
                        [
                            sampleName:   "SRR1066657",
                            sampleNumber: "3",
                            lane:         "002",
                            readType:     "single"
                        ],
                        [
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR1066657_S3_L002_R1_001.fastq.gz")
                        ]
                    ]
                )
                """
            }
        }

        then {
            // test workflow status
            assert workflow.success
            assert workflow.trace.succeeded().size() == 2

            // test read trimming and concatenation
            with(workflow.out.reads_pre_align) {
                assert size() == 1
                with(get(0)) {
                    assert size() == 2
                    assert get(1) ==~ /^.*\/SRR1066657_trimmed_R1.fastq.gz/
                }
            }
            assert snapshot(workflow.out.reads_pre_align).match()

            // test read trimming log
            with(workflow.out.trim_log) {
                assert size() == 1
                assert get(0) ==~ /^.*\/SRR1066657_cutadapt-log.txt/
            }
        }

    }

    test("PROCESS_READS concatenates and trims reads by default -- PE reads.") {

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [
                            sampleName:   "SRR6924569",
                            sampleNumber: "1",
                            lane:         "001",
                            readType:     "paired"
                        ],
                        [
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L001_R1_001.fastq.gz"),
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L001_R2_001.fastq.gz")
                        ]
                    ],
                    [
                        [
                            sampleName:   "SRR6924569",
                            sampleNumber: "1",
                            lane:         "002",
                            readType:     "paired"
                        ],
                        [
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L002_R1_001.fastq.gz"),
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L002_R2_001.fastq.gz")
                        ]
                    ]
                )
                """
            }
        }

        then {
            // test workflow status
            assert workflow.success
            assert workflow.trace.succeeded().size() == 2

            // test read trimming and concatenation
            with(workflow.out.reads_pre_align) {
                assert size() == 1
                with(get(0)) {
                    assert size() == 2
                    assert get(1).get(0) ==~ /^.*\/SRR6924569_trimmed_R1.fastq.gz/
                    assert get(1).get(1) ==~ /^.*\/SRR6924569_trimmed_R2.fastq.gz/
                }
            }
            assert snapshot(workflow.out.reads_pre_align).match()

            // test read trimming log
            with(workflow.out.trim_log) {
                assert size() == 1
                assert get(0) ==~ /^.*\/SRR6924569_cutadapt-log.txt/
            }
        }

    }

    test("PROCESS_READS only concatenates reads when `params.skipTrimReads = true` -- SE reads.") {

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
                skipTrimReads     = true
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [
                            sampleName:   "SRR1066657",
                            sampleNumber: "3",
                            lane:         "001",
                            readType:     "single"
                        ],
                        [
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR1066657_S3_L001_R1_001.fastq.gz")
                        ]
                    ],
                    [
                        [
                            sampleName:   "SRR1066657",
                            sampleNumber: "3",
                            lane:         "002",
                            readType:     "single"
                        ],
                        [
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR1066657_S3_L002_R1_001.fastq.gz")
                        ]
                    ]
                )
                """
            }
        }

        then {
            // test workflow status
            assert workflow.success
            assert workflow.trace.succeeded().size() == 1

            // test read trimming and concatenation
            with(workflow.out.reads_pre_align) {
                assert size() == 1
                with(get(0)) {
                    assert size() == 2
                    assert get(1) ==~ /^.*\/SRR1066657_R1.fastq.gz/
                }
            }
            assert snapshot(workflow.out.reads_pre_align).match()

            // test read trimming log
            with(workflow.out.trim_log) {
                assert get(0) == 'EMPTY'
            }
        }

    }

}
