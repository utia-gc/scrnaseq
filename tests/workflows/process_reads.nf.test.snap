{
    "PROCESS_READS concatenates and trims reads by default -- PE reads.": {
        "content": [
            [
                [
                    {
                        "sampleName": "SRR6924569",
                        "readType": "paired"
                    },
                    [
                        "SRR6924569_trimmed_R1.fastq.gz:md5,68b0217aba5b177e752fd689d251527f",
                        "SRR6924569_trimmed_R2.fastq.gz:md5,6477b103ff26f0e32294ab80b265b184"
                    ]
                ]
            ]
        ],
        "timestamp": "2023-08-21T15:28:46+0000"
    },
    "PROCESS_READS only concatenates reads when `params.skipTrimReads = true` -- SE reads.": {
        "content": [
            [
                [
                    {
                        "sampleName": "SRR1066657",
                        "readType": "single"
                    },
                    "SRR1066657_R1.fastq.gz:md5,ccabf5ada78353bc8c0584940ab92d35"
                ]
            ]
        ],
        "timestamp": "2023-08-21T15:28:46+0000"
    },
    "PROCESS_READS concatenates and trims reads by default -- SE reads.": {
        "content": [
            [
                [
                    {
                        "sampleName": "SRR1066657",
                        "readType": "single"
                    },
                    "SRR1066657_trimmed_R1.fastq.gz:md5,98f53500019dd1bdff84a8255e5bf975"
                ]
            ]
        ],
        "timestamp": "2023-08-21T15:28:46+0000"
    }
}